lib/internal/streams/add-abort-signal.js

```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+const {
+  AbortError,
+  codes,
+} = require('internal/errors');
+
+const { ERR_INVALID_ARG_TYPE } = codes;
+
+const {
+  validateAbortSignal,
+} = require('internal/validators');
+
+function isReadable(obj) {
+  return !!(obj && typeof obj.pipe === 'function');
+}
+
+module.exports = function addAbortSignal(signal, readable) {
+  validateAbortSignal(signal, 'signal');
+  if (!isReadable(readable)) {
+    throw new ERR_INVALID_ARG_TYPE('readable', 'stream.Readable', readable);
+  }
+  const onAbort = () => {
+    readable.destroy(new AbortError());
+  };
+  if (signal.aborted) {
+    onAbort();
+  } else {
+    signal.addEventListener('abort', onAbort);
+    readable.on('close', () => signal.removeEventListener('abort', onAbort));
+  }
+  return readable;
```

Comment:
Return readable here seems unnecessary. Would be better to return a cleanup function?

Reply:
I think it really helps with ergonomics to return the stream here (see the tests for why). No very strong feelings though.

Reply:
That's fine but then we need a `removeAbortSignal` function.

Reply:
Why?

The case is:

 - I have a stream
 - I have a signal
 - I want the stream destroyed when I abort the controller linked to the signal

Why would I want to remove the listener from the stream? Is the use case for `removeAbortSignal` that the stream's lifespan is much longer than the signal's?

Reply:
Because you might run a function multiple times that does something and then changes it mind.


```js
function foo (stream) {
  const signal = ...
  stream.addAbortSignal(signal)
  readStreamUntilNextNewLine(stream)
  stream.removeAbortSignal(signal)
}

// Call foo in loop

```

Reply:
Why would you want to destroy the entire stream because you regret reading a single line? That sounds... odd no?

Reply:
If you can dynamically add an abort signal I would expect it to be possible to dynamically remove it.

Reply:
From a theoretical point of view I see it and from an implementation PoV this is easy to add but I just don't see any use case where people would use this.

I will happily concede with a single real use case

Reply:
> Why would you want to destroy the entire stream because you regret reading a single line? That sounds... odd no?

It could be a e.g. a timeout.

Reply:
But that's not what it would timeout - it would destroy the stream not cancel the request for data.

Reply:
Yes? 

BTW. I won't block because of this.

Reply:
I wouldn't implement it using this but I can see how e.g. it could be used for implementing a upstream retry mechanism where a remove function would be necessary to avoid memory leaks.

Reply:
As this is a month old already - I hope it's clear by now I am less concerned with landing this quickly and more with landing this right. I expected discussion (which is why this is deliberately kinda small)

Worth saying (_you_ know I'm on your side given I pushed for you in core in the first place but there might be more readers here) that I appreciate the feedback and discussion, there are maybe 10 people total in this discussion affecting APIs that impact millions of users.

> I wouldn't implement it using this but I can see how e.g. it could be used for implementing a upstream retry mechanism where a remove function would be necessary to avoid memory leaks.

A code example would be appreciated if you're up to make one. I honestly don't understand the use case. 

I tried (not for too long admittedly, like 20m) to come up with use cases for `removeAbortSignal` and couldn't find any that made sense.

I don't understand why I'd destroy the whole stream for "regretting" reading a line and in that case the feature request IIRC is passing a signal to "`read`" (in quotes since well, that wouldn't be possible/good given buffering, watermarks and read being sync and only consuming already ready data) and not to the stream itself.

Reply:
> there are maybe 10 people total in this discussion affecting APIs that impact millions of users.

That bit (in streams in particular) is always kind of terrifying ^^

Reply:
This is the best I can come up with at the moment:

```js
async function* read (signal, { start, end }) {
  const src = findInCacheOrCreateStream({ start }) // e.g. a cache of open range HTTP request
  src.addAbortSignal(signal)
  yield* readUntil(src, { end })
  if (!src.readableEnded) {
    // src can be used by other readers starting at 'end'.
    cacheStream(src, { start: end })
    src.removeAbortSignal(signal) 
  }
}
```

Reply:
It's a good example thanks.

I'm not sure I understand why I would want to destroy the entire cached stream because the user who called `read` decided to abort. That is, the thing we really want to abort here is the `readUntil` - and that would likely be need to still complete a single "read" rather than abort the read "in flight" so we know what `end` is in order to cache the stream.

I'll iterate a bit on the example and see if I can come up with a version where I want to destroy the whole stream (that is, I no longer want it to be in the cache for some reason?).

Reply:
> I'm not sure I understand why I would want to destroy the entire cached stream because the user who called read decided to abort.

Yea, good point.

I don't really have a better example at the moment.